### ISSUE TITLE
Potential performance regression in the `*IterateForEach*Span*` benchmarks

### ISSUE THREAD

**Issue by adamsitnik** [dotnet/runtime#55416]
While doing the .NET 5 vs .NET 6 preview 7 manual perf run I've found something that seems to be ARM64, ARM and AMD (x64) specific regression.

Repro:

```cmd
py .\scripts\benchmarks_ci.py -f net5.0 net6.0 --filter "System.Collections.IterateForEach<int32>.ReadOnlySpan" --bdn-arguments "--disasm true --memoryRandomization true"
```

Results from multiple machines:

<details>

## System.Collections.IterateForEach<String>.ReadOnlySpan(Size: 512)

| Result |   Base |    Diff | Ratio | Alloc Delta | Modality | Operating System        | Bit   | Processor Name                                | Base V        | Diff V      |
| ------ | ------:| -------:| -----:| -----------:| -------- | ----------------------- | ----- | --------------------------------------------- | ------------- | ------------ |
| Slower | 125.03 | 1103.94 |  0.11 |          +0 |          | Windows 10.0.19041.1052 | X64   | AMD Ryzen Threadripper PRO 3945WX 12-Cores    | 5.0.721.25508 | 6.0.21.35216|
| Same   | 216.85 |  219.31 |  0.99 |          +0 |          | Windows 10.0.18363.1621 | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz             | 5.0.721.25508 | 6.0.21.32901|
| Same   | 241.37 |  241.29 |  1.00 |          +0 |          | Windows 10.0.19043.1052 | X64   | Intel Core i5-8500 CPU 3.00GHz (Coffee Lake)  | 5.0.721.25508 | 6.0.21.32802|
| Same   | 241.67 |  242.53 |  1.00 |          +0 |          | Windows 10.0.19042.685  | X64   | Intel Core i7-5557U CPU 3.10GHz (Broadwell)   | 5.0.721.25508 | 6.0.21.32901|
| Same   | 279.40 |  284.73 |  0.98 |          +0 | bimodal  | Windows 10.0.19043.1052 | X64   | Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R) | 5.0.721.25508 | 6.0.21.32802|
| Same   | 320.13 |  311.72 |  1.03 |          +0 | several? | Windows 10.0.22000      | X64   | Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R) | 5.0.721.25508 | 6.0.21.32901|
| Slower | 189.81 |  215.37 |  0.88 |          +0 |          | Windows 10.0.22000      | X64   | Intel Core i7-8700 CPU 3.20GHz (Coffee Lake)  | 5.0.721.25508 | 6.0.21.32802|
| Same   | 407.89 |  394.83 |  1.03 |          +0 |          | Windows 7 SP1           | X64   | Intel Core2 Duo CPU T9600 2.80GHz             | 5.0.721.25508 | 6.0.21.32901|
| Same   | 204.40 |  200.56 |  1.02 |          +0 |          | ubuntu 18.04            | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz             | 5.0.721.25508 | 6.0.21.32901|
| Slower | 403.53 | 5274.27 |  0.08 |          +0 |          | ubuntu 16.04            | Arm64 | Unknown processor                             | 5.0.721.25508 | 6.0.21.35701|
| Same   | 352.76 |  352.49 |  1.00 |          +0 |          | Windows 10.0.19043.1052 | Arm64 | Microsoft SQ1 3.0 GHz                         | 5.0.721.25508 | 6.0.21.32901|
| Same   | 140.73 |  139.72 |  1.01 |          +0 |          | Windows 10.0.18363.1621 | X86   | Intel Xeon CPU E5-1650 v4 3.60GHz             | 5.0.721.25508 | 6.0.21.32901|
| Slower | 329.66 | 1394.41 |  0.24 |          +0 |          | Windows 10.0.19043.1052 | Arm   | Microsoft SQ1 3.0 GHz                         | 5.0.721.25508 | 6.0.21.35216|
| Same   | 331.44 |  330.27 |  1.00 |          +0 |          | macOS Big Sur 11.3.1    | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)     | 5.0.721.25508 | 6.0.21.32901|
| Same   | 323.04 |  322.33 |  1.00 |          +0 |          | macOS Big Sur 11.4      | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)     | 5.0.721.25508 | 6.0.21.32901|
| Same   | 239.51 |  238.36 |  1.00 |          +0 |          | macOS Big Sur 11.4      | X64   | Intel Core i7-5557U CPU 3.10GHz (Broadwell)   | 5.0.721.25508 | 6.0.21.32901|

## System.Collections.IterateForEach<Int32>.ReadOnlySpan(Size: 512)

| Result |   Base |    Diff | Ratio | Alloc Delta | Modality | Operating System        | Bit   | Processor Name                                | Base V        | Diff V      |
| ------ | ------:| -------:| -----:| -----------:| --------:| ----------------------- | ----- | --------------------------------------------- | ------------- | ------------ |
| Slower | 123.26 | 1108.25 |  0.11 |          +0 |          | Windows 10.0.19041.1052 | X64   | AMD Ryzen Threadripper PRO 3945WX 12-Cores    | 5.0.721.25508 | 6.0.21.35216|
| Same   | 208.98 |  203.03 |  1.03 |          +0 |          | Windows 10.0.18363.1621 | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz             | 5.0.721.25508 | 6.0.21.32901|
| Same   | 166.68 |  160.38 |  1.04 |          +0 |          | Windows 10.0.19043.1052 | X64   | Intel Core i5-8500 CPU 3.00GHz (Coffee Lake)  | 5.0.721.25508 | 6.0.21.32802|
| Same   | 233.36 |  229.13 |  1.02 |          +0 |          | Windows 10.0.19042.685  | X64   | Intel Core i7-5557U CPU 3.10GHz (Broadwell)   | 5.0.721.25508 | 6.0.21.32901|
| Slower | 250.30 |  282.27 |  0.89 |          +0 |          | Windows 10.0.19043.1052 | X64   | Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R) | 5.0.721.25508 | 6.0.21.32802|
| Slower | 251.59 |  285.56 |  0.88 |          +0 |          | Windows 10.0.22000      | X64   | Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R) | 5.0.721.25508 | 6.0.21.32901|
| Same   | 192.04 |  213.50 |  0.90 |          +0 |          | Windows 10.0.22000      | X64   | Intel Core i7-8700 CPU 3.20GHz (Coffee Lake)  | 5.0.721.25508 | 6.0.21.32802|
| Same   | 403.58 |  405.14 |  1.00 |          +0 |          | Windows 7 SP1           | X64   | Intel Core2 Duo CPU T9600 2.80GHz             | 5.0.721.25508 | 6.0.21.32901|
| Same   | 196.03 |  199.04 |  0.98 |          +0 |          | ubuntu 18.04            | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz             | 5.0.721.25508 | 6.0.21.32901|
| Slower | 395.46 | 4943.41 |  0.08 |          +0 |          | ubuntu 16.04            | Arm64 | Unknown processor                             | 5.0.721.25508 | 6.0.21.35701|
| Same   | 346.01 |  332.33 |  1.04 |          +0 |          | Windows 10.0.19043.1052 | Arm64 | Microsoft SQ1 3.0 GHz                         | 5.0.721.25508 | 6.0.21.32901|
| Same   | 143.67 |  143.86 |  1.00 |          +0 |          | Windows 10.0.18363.1621 | X86   | Intel Xeon CPU E5-1650 v4 3.60GHz             | 5.0.721.25508 | 6.0.21.32901|
| Slower | 346.26 | 1362.83 |  0.25 |          +0 |          | Windows 10.0.19043.1052 | Arm   | Microsoft SQ1 3.0 GHz                         | 5.0.721.25508 | 6.0.21.35216|
| Same   | 326.15 |  324.09 |  1.01 |          +0 |          | macOS Big Sur 11.3.1    | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)     | 5.0.721.25508 | 6.0.21.32901|
| Same   | 318.47 |  315.87 |  1.01 |          +0 |          | macOS Big Sur 11.4      | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)     | 5.0.721.25508 | 6.0.21.32901|
| Same   | 230.46 |  228.78 |  1.01 |          +0 |          | macOS Big Sur 11.4      | X64   | Intel Core i7-5557U CPU 3.10GHz (Broadwell)   | 5.0.721.25508 | 6.0.21.32901|

## System.Collections.IterateForEach<String>.Span(Size: 512)

| Result |   Base |    Diff | Ratio | Alloc Delta | Modality | Operating System        | Bit   | Processor Name                                | Base V        | Diff V      |
| ------ | ------:| -------:| -----:| -----------:| --------:| ----------------------- | ----- | --------------------------------------------- | ------------- | ------------ |
| Slower | 127.87 | 1111.12 |  0.12 |          +0 |          | Windows 10.0.19041.1052 | X64   | AMD Ryzen Threadripper PRO 3945WX 12-Cores    | 5.0.721.25508 | 6.0.21.35216|
| Same   | 216.69 |  210.35 |  1.03 |          +0 |          | Windows 10.0.18363.1621 | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz             | 5.0.721.25508 | 6.0.21.32901|
| Slower | 186.30 |  241.93 |  0.77 |          +0 |          | Windows 10.0.19043.1052 | X64   | Intel Core i5-8500 CPU 3.00GHz (Coffee Lake)  | 5.0.721.25508 | 6.0.21.32802|
| Same   | 241.10 |  234.05 |  1.03 |          +0 |          | Windows 10.0.19042.685  | X64   | Intel Core i7-5557U CPU 3.10GHz (Broadwell)   | 5.0.721.25508 | 6.0.21.32901|
| Same   | 275.56 |  269.14 |  1.02 |          +0 |          | Windows 10.0.19043.1052 | X64   | Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R) | 5.0.721.25508 | 6.0.21.32802|
| Faster | 299.24 |  190.93 |  1.57 |          +0 |          | Windows 10.0.22000      | X64   | Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R) | 5.0.721.25508 | 6.0.21.32901|
| Slower | 183.80 |  207.79 |  0.88 |          +0 |          | Windows 10.0.22000      | X64   | Intel Core i7-8700 CPU 3.20GHz (Coffee Lake)  | 5.0.721.25508 | 6.0.21.32802|
| Same   | 410.40 |  405.76 |  1.01 |          +0 |          | Windows 7 SP1           | X64   | Intel Core2 Duo CPU T9600 2.80GHz             | 5.0.721.25508 | 6.0.21.32901|
| Same   | 207.41 |  206.67 |  1.00 |          +0 |          | ubuntu 18.04            | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz             | 5.0.721.25508 | 6.0.21.32901|
| Slower | 418.07 | 5286.05 |  0.08 |          +0 |          | ubuntu 16.04            | Arm64 | Unknown processor                             | 5.0.721.25508 | 6.0.21.35701|
| Same   | 339.92 |  360.51 |  0.94 |          +0 |          | Windows 10.0.19043.1052 | Arm64 | Microsoft SQ1 3.0 GHz                         | 5.0.721.25508 | 6.0.21.32901|
| Same   | 151.40 |  150.49 |  1.01 |          +0 |          | Windows 10.0.18363.1621 | X86   | Intel Xeon CPU E5-1650 v4 3.60GHz             | 5.0.721.25508 | 6.0.21.32901|
| Slower | 332.07 | 1380.42 |  0.24 |          +0 |          | Windows 10.0.19043.1052 | Arm   | Microsoft SQ1 3.0 GHz                         | 5.0.721.25508 | 6.0.21.35216|
| Same   | 335.51 |  334.64 |  1.00 |          +0 |          | macOS Big Sur 11.3.1    | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)     | 5.0.721.25508 | 6.0.21.32901|
| Same   | 327.04 |  327.25 |  1.00 |          +0 |          | macOS Big Sur 11.4      | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)     | 5.0.721.25508 | 6.0.21.32901|
| Same   | 241.14 |  240.91 |  1.00 |          +0 |          | macOS Big Sur 11.4      | X64   | Intel Core i7-5557U CPU 3.10GHz (Broadwell)   | 5.0.721.25508 | 6.0.21.32901|

## System.Collections.IterateForEach<Int32>.Span(Size: 512)

| Result |   Base |    Diff | Ratio | Alloc Delta | Modality | Operating System        | Bit   | Processor Name                                | Base V        | Diff V      |
| ------ | ------:| -------:| -----:| -----------:| -------- | ----------------------- | ----- | --------------------------------------------- | ------------- | ------------ |
| Slower | 123.08 | 1109.80 |  0.11 |          +0 |          | Windows 10.0.19041.1052 | X64   | AMD Ryzen Threadripper PRO 3945WX 12-Cores    | 5.0.721.25508 | 6.0.21.35216|
| Same   | 208.99 |  202.05 |  1.03 |          +0 |          | Windows 10.0.18363.1621 | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz             | 5.0.721.25508 | 6.0.21.32901|
| Same   | 168.16 |  159.26 |  1.06 |          +0 |          | Windows 10.0.19043.1052 | X64   | Intel Core i5-8500 CPU 3.00GHz (Coffee Lake)  | 5.0.721.25508 | 6.0.21.32802|
| Same   | 233.38 |  229.84 |  1.02 |          +0 |          | Windows 10.0.19042.685  | X64   | Intel Core i7-5557U CPU 3.10GHz (Broadwell)   | 5.0.721.25508 | 6.0.21.32901|
| Same   | 252.07 |  283.22 |  0.89 |          +0 | bimodal  | Windows 10.0.19043.1052 | X64   | Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R) | 5.0.721.25508 | 6.0.21.32802|
| Slower | 246.32 |  283.00 |  0.87 |          +0 |          | Windows 10.0.22000      | X64   | Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R) | 5.0.721.25508 | 6.0.21.32901|
| Same   | 209.99 |  212.25 |  0.99 |          +0 |          | Windows 10.0.22000      | X64   | Intel Core i7-8700 CPU 3.20GHz (Coffee Lake)  | 5.0.721.25508 | 6.0.21.32802|
| Same   | 403.82 |  405.26 |  1.00 |          +0 |          | Windows 7 SP1           | X64   | Intel Core2 Duo CPU T9600 2.80GHz             | 5.0.721.25508 | 6.0.21.32901|
| Same   | 196.18 |  199.00 |  0.99 |          +0 |          | ubuntu 18.04            | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz             | 5.0.721.25508 | 6.0.21.32901|
| Slower | 395.45 | 5200.84 |  0.08 |          +0 |          | ubuntu 16.04            | Arm64 | Unknown processor                             | 5.0.721.25508 | 6.0.21.35701|
| Same   | 332.26 |  345.87 |  0.96 |          +0 |          | Windows 10.0.19043.1052 | Arm64 | Microsoft SQ1 3.0 GHz                         | 5.0.721.25508 | 6.0.21.32901|
| Same   | 143.98 |  143.81 |  1.00 |          +0 |          | Windows 10.0.18363.1621 | X86   | Intel Xeon CPU E5-1650 v4 3.60GHz             | 5.0.721.25508 | 6.0.21.32901|
| Slower | 346.31 | 1363.60 |  0.25 |          +0 |          | Windows 10.0.19043.1052 | Arm   | Microsoft SQ1 3.0 GHz                         | 5.0.721.25508 | 6.0.21.35216|
| Same   | 326.52 |  323.78 |  1.01 |          +0 |          | macOS Big Sur 11.3.1    | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)     | 5.0.721.25508 | 6.0.21.32901|
| Same   | 318.34 |  315.92 |  1.01 |          +0 |          | macOS Big Sur 11.4      | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)     | 5.0.721.25508 | 6.0.21.32901|
| Same   | 230.54 |  228.77 |  1.01 |          +0 |          | macOS Big Sur 11.4      | X64   | Intel Core i7-5557U CPU 3.10GHz (Broadwell)   | 5.0.721.25508 | 6.0.21.32901|

</details>

Disassembly from my AMD machine:

<details>

## .NET 5.0.7 (5.0.721.25508), X64 RyuJIT
```assembly
; System.Collections.IterateForEach`1[[System.Int32, System.Private.CoreLib]].ReadOnlySpan()
       xor       eax,eax
       mov       rdx,[rcx+8]
       test      rdx,rdx
       jne       short M00_L00
       xor       ecx,ecx
       xor       r8d,r8d
       jmp       short M00_L01
M00_L00:
       lea       rcx,[rdx+10]
       mov       r8d,[rdx+8]
M00_L01:
       xor       edx,edx
       test      r8d,r8d
       jle       short M00_L03
M00_L02:
       movsxd    rax,edx
       mov       eax,[rcx+rax*4]
       inc       edx
       cmp       edx,r8d
       jl        short M00_L02
M00_L03:
       ret
; Total bytes of code 47
```

## .NET 6.0.0 (6.0.21.35701), X64 RyuJIT
```assembly
; System.Collections.IterateForEach`1[[System.Int32, System.Private.CoreLib]].ReadOnlySpan()
       sub       rsp,58
       vzeroupper
       vxorps    xmm4,xmm4,xmm4
       vmovdqa   xmmword ptr [rsp+30],xmm4
       vmovdqa   xmmword ptr [rsp+40],xmm4
       xor       eax,eax
       mov       [rsp+54],eax
       mov       rax,[rcx+8]
       test      rax,rax
       jne       short M00_L00
       vxorps    xmm0,xmm0,xmm0
       vmovdqu   xmmword ptr [rsp+40],xmm0
       jmp       short M00_L01
M00_L00:
       lea       rdx,[rax+10]
       mov       [rsp+40],rdx
       mov       eax,[rax+8]
       mov       [rsp+48],eax
M00_L01:
       mov       rax,[rsp+40]
       mov       [rsp+30],rax
       mov       eax,[rsp+48]
       mov       [rsp+38],eax
       xor       eax,eax
       mov       [rsp+2C],eax
       cmp       dword ptr [rsp+38],0
       jle       short M00_L03
M00_L02:
       mov       eax,[rsp+38]
       cmp       [rsp+2C],eax
       jae       short M00_L04
       mov       rax,[rsp+30]
       mov       edx,[rsp+2C]
       movsxd    rdx,edx
       mov       eax,[rax+rdx*4]
       mov       [rsp+28],eax
       mov       eax,[rsp+28]
       mov       [rsp+54],eax
       mov       eax,[rsp+2C]
       inc       eax
       mov       [rsp+2C],eax
       mov       eax,[rsp+2C]
       cmp       eax,[rsp+38]
       jl        short M00_L02
M00_L03:
       mov       eax,[rsp+54]
       add       rsp,58
       ret
M00_L04:
       call      CORINFO_HELP_RNGCHKFAIL
       int       3
; Total bytes of code 169
```

</details>

cc @kunalspathak @AndyAyersMS @DrewScoggins

**Comment by SingleAccretion** [comment]
The .NET 6 assembly looks suspiciously like Debug codegen.

Could this be #55300?

**Comment by adamsitnik** [comment]
By looking at the historical data it seems that it might already be gone:

https://pvscmdupload.blob.core.windows.net/reports/allTestHistory%2frefs%2fheads%2fmain_x64_Windows%2010.0.19042%2famd%2fSystem.Collections.IterateForEach(Int32).Span(Size%3a%20512).html

![image](https://user-images.githubusercontent.com/6011991/125116885-ae3c7200-e0ed-11eb-9a5c-65b1f4d9b0da.png)

but it would be great to be 100% sure and understand where it was coming from

**Comment by DrewScoggins** [comment]
I think this was fixed by this, https://github.com/dotnet/runtime/pull/55300

**Comment by kunalspathak** [comment]
@adamsitnik - you might see lot of regressions if your changes don't include #55300. It might be good to conduct the experiments with #55300 included.

**Comment by adamsitnik** [comment]
Since it seems that the regression was solved by #55300, I am closing the issue.

> you might see lot of regressions

you are right, it's everywhere ;)
